node/rpc/src/rpc_ws_handler.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_std::sync::{Arc, Mutex};
+use crossbeam::atomic::AtomicCell;
+use futures::{SinkExt, StreamExt};
+use log::{error, info, warn};
+use tide_websockets::{Message, WebSocketConnection};
+
+use beacon::Beacon;
+use blockstore::BlockStore;
+use wallet::KeyStore;
+
+use crate::chain_api::Subscription;
+use crate::data_types::{JsonRpcServerState, StreamingData};
+use crate::rpc_util::{
+    get_error_str, get_rpc_call_response, get_rpc_call_result, RPC_METHOD_CHAIN_HEAD_SUB,
+    RPC_METHOD_CHAIN_NOTIFY,
+};
+use chain::headchange_json::HeadChangeJson;
+
+pub async fn rpc_ws_handler<DB, KS, B>(
+    request: tide::Request<JsonRpcServerState>,
+    ws_stream: WebSocketConnection,
+) -> Result<(), tide::Error>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    B: Beacon + Send + Sync + 'static,
+{
+    let rpc_server = request.state();
+    let (ws_sender, mut ws_receiver) = ws_stream.split();
+    let ws_sender = Arc::new(Mutex::new(ws_sender));
+    let socket_active = Arc::new(AtomicCell::new(true));
+
+    info!("Accepted WS connection!");
+
+    while let Some(message_result) = ws_receiver.next().await {
+        match message_result {
+            Ok(message) => {
+                let request_text = message.into_text()?;
+
+                if !request_text.is_empty() {
+                    info!("RPC Request Received: {:?}", &request_text);
+
+                    match serde_json::from_str(&request_text)
+                        as Result<jsonrpc_v2::RequestObject, serde_json::Error>
+                    {
+                        Ok(call) => match &*call.method_ref() {
+                            RPC_METHOD_CHAIN_NOTIFY => {
+                                let Subscription { subscription_id } = get_rpc_call_result(
+                                    rpc_server.clone(),
+                                    jsonrpc_v2::RequestObject::request()
+                                        .with_method(RPC_METHOD_CHAIN_HEAD_SUB)
+                                        .finish(),
+                                )
+                                .await?;
+
+                                info!(
+                                    "RPC WS ChainNotify for subscription ID: {}",
+                                    subscription_id
+                                );
+
```

Comment:
When I call chainnotify, i expect the first response to be the channel id, and the second response to be a "current" headchange event. Only after those two we start streaming the incoming tipsets out.

Reply:
I don't quite understand this instruction, you mean there should be two separate JSON responses immediately after calling ChainNotify over the websocket?

Reply:
Alright man. I'll take a look at how it worked before.

Reply:
Dope. Its easy to miss and should be an easy fix 👍

Reply:
For reference, this is how we do it currently in main: 
* Send the Channel ID (https://github.com/ChainSafe/forest/blob/main/node/rpc/src/lib.rs#L448)
* sub_head_changes will return the `current` event, if you look at it.  (https://github.com/ChainSafe/forest/blob/main/node/rpc/src/lib.rs#L435)